lib/modules/home/home_interactor.dart

```diff
@@ -1,7 +1,70 @@
 part of 'home_module.dart';
 
-abstract class HomeInteractorDelegate {}
+abstract class HomeInteractorDelegate {
+  BehaviorSubject<List<SurveyInfo>> get surveysDidFetchFromCached;
 
-abstract class HomeInteractor extends Interactor<HomeInteractorDelegate> {}
+  BehaviorSubject<Exception> get surveysDidFailToFetchFromCached;
 
-class HomeInteractorImpl extends HomeInteractor {}
+  BehaviorSubject<Tuple2<List<SurveyInfo>, bool>> get surveysDidFetchFromRemote;
+
+  BehaviorSubject<Exception> get surveysDidFailToFetchFromRemote;
+
+  BehaviorSubject<UserInfo> get authenticatedUserDidGet;
+}
+
+abstract class HomeInteractor extends Interactor<HomeInteractorDelegate> {
+  void fetchSurveysFromCached();
+
+  void fetchSurveysFromRemote();
+
+  void getAuthenticatedUser();
+}
+
+class HomeInteractorImpl extends HomeInteractor {
+  final SurveyRepository _surveyRepository = locator.get();
+  final AuthRepository _authRepository = locator.get();
+
+  @override
+  void fetchSurveysFromCached() {
+    _surveyRepository
+        .fetchSurveysFromCached()
+        .then((value) => delegate?.surveysDidFetchFromCached.add(value))
+        .onError<Exception>((exception, _) =>
+            delegate?.surveysDidFailToFetchFromCached.add(exception));
+  }
+
+  @override
+  void fetchSurveysFromRemote() {
+    Future.wait([
+      _surveyRepository.fetchSurveysFromCached(),
+      _surveyRepository.fetchSurveysFromRemote()
+    ]).then((values) {
+      final cachedSurveys = values[0];
+      final remoteSurveys = values[1];
+
+      final hasChanges = _isCachedSurveysChanged(cachedSurveys, remoteSurveys);
+      delegate?.surveysDidFetchFromRemote
+          .add(Tuple2(remoteSurveys, hasChanges));
+    }).onError<Exception>((exception, _) {
+      delegate?.surveysDidFailToFetchFromRemote.add(exception);
+    });
+  }
+
+  @override
+  void getAuthenticatedUser() {
+    delegate?.authenticatedUserDidGet.add(_authRepository.user!);
+  }
+
+  bool _isCachedSurveysChanged(
+      List<SurveyInfo> cachedSurveys, List<SurveyInfo> remoteSurveys) {
+    if (cachedSurveys.length != remoteSurveys.length) return true;
+
+    int count = 0;
+    for (final survey in cachedSurveys) {
+      if (!remoteSurveys.contains(survey)) return false;
```

Comment:
hmm I'm a bit suspicious of this condition 🤔  because you might potential return false when the `remoteSurveys` has a different set of data than the `cachedSurveys` -> this value is later used here:

```
if (hasChanges) view.showSurveys(surveys);
```

which means if hashChanges == false, then you will not call `showSurveys` - which is from remote, the updated one?

Reply:
@sleepylee my bad, it should return `true` here 😓